Algorithm/ThreadManager.py

```diff
@@ -5,36 +5,29 @@
 # Ronen Rozen - 203024542
 ##############################
 
-from contextlib import contextmanager
-from threading import Thread, Lock
-from typing import Tuple
 from collections import namedtuple
+from threading import Thread
+from typing import Tuple
 
 from Algorithm.Model import learning_model
 from Utils.Log import writer
 
 TIMEOUT_THREAD = 10  # seconds
 
 
-@contextmanager
```

Comment:
How do we handle with dead locks?

Reply:
there are no deadlocks, each train is independent on others.